### PR TITLE
expression: fix incorrect result when the result of casting const value to duration type is null

### LINKED ISSUE #51842: Incorrect query reuslt  caused by NULLIF keyword
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
Consider the following statements. 

In theory, query1 and query2 should return the same result. However, query1 return nothing `NULL` while query2 return nothing.

```
CREATE TABLE t0(c0 DOUBLE);
REPLACE INTO t0(c0) VALUES (0.40194983109852933);
CREATE VIEW v0(c0) AS SELECT CAST(')' AS TIME) FROM t0 WHERE '0.030417148673465677';

mysql> SELECT f1 FROM (SELECT NULLIF(v0.c0, 1371581446) AS f1 FROM v0, t0) AS t WHERE f1 <=> 1292367147; --query1
+------+
| f1   |
+------+
| NULL |
+------+
1 row in set, 3 warnings (0.01 sec)

mysql> SELECT f1 FROM (SELECT NULLIF(v0.c0, 1371581446) AS f1, (NULLIF(v0.c0, 1371581446) <=> 1292367147 ) IS TRUE AS flag FROM v0, t0) AS t WHERE flag=1; --query2
Empty set, 3 warnings (0.00 sec)
```

### 2. What is your TiDB version? (Required)
v7.6.0, v6.4.0

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #51842

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
fix incorrect result when the result of casting const value to duration type is null
```

### CODE REVIEW

File: pkg/expression/builtin_compare.go [hunk 1]
```diff
@@ -1586,6 +1602,28 @@ func (c *compareFunctionClass) refineArgs(ctx BuildContext, args []Expression) (
...
+	if c.op == opcode.NullEQ && arg1IsCon && arg1.DeferredExpr == nil && !arg0IsCon && arg0Type.GetType() == mysql.TypeDuration {
+		isNull, err := castToDurationIsNull(ctx, arg1)
+		if err != nil {
+			return nil, err
+		}
+
+		if isNull {
+			return []Expression{NewZero(), NewOne()}, nil
+		}
+	}
```
Comment (winoros) [inline pkg/expression/builtin_compare.go:1625]: Add some comments for why we're add this if block

Comment (xzhangxian1008) [reply pkg/expression/builtin_compare.go:1630]: > Add some comments for why we're add this if block

done

**Review by XuHuaiyu — APPROVED** [review]

**Review by winoros — APPROVED** [review]

**Pull request merged**